`grpcio-proto` 0.7 and 0.8 now fail to build as a dependency

**Describe the bug**
`grpcio-proto` 0.8 brings in `protobuf-build`. `protobuf-build` versions 12.1+  depend on `grpcio-compiler>=0.8.0`, which as of today brings in `grpcio-compiler` 0.9.0, which causes the build to fail when building `grpc-proto` 0.8 as a dependency.

`grpcio-proto` 0.7 (and maybe older versions) has the same issue, because `protobuf-build` 0.11.4 brings in `grpcio-compiler>=0.5.0`.

This can be worked around with cargo patches, but it is a bit painful to do so.

**To Reproduce**
Steps to reproduce the behavior:
1. Make a new crate and add `grpcio-proto = "0.8"` to the dependencies array in its `Cargo.toml`.
2. `cargo check`

```
    Checking grpcio-proto v0.8.1
error[E0433]: failed to resolve: could not find `unimplemented_call` in `grpcio`
   --> /home/matthew/misc/grpc-test/target/debug/build/grpcio-proto-ca27f1d4f4011d16/out/testing/test_grpc.rs:189:17
    |
189 |         grpcio::unimplemented_call!(ctx, sink)
    |                 ^^^^^^^^^^^^^^^^^^ could not find `unimplemented_call` in `grpcio`
<many copies of the same error...>
```

**Expected behavior**
Old versions of `grpcio-protobuf` should continue to build as dependencies without cargo-patching inner dependencies when a new version is released. Unfortunately, I think this can only be done by yanking `grpcio-compiler` 0.9.0 and/or releasing a version of it that works with old `grpcio-proto` versions.

Alternatively, new patch releases of `protobuf-build` and/or old versions of `grpcio-proto` to fix the issue might help, but I haven't fully thought through how that would need to work -- I understand there's a cyclic dependency the `>=` dependency fixes (at the cost of breaking old versions' builds).

**System information**
* CPU architecture: x86_64
* Distribution and kernel version: Fedora 33, 5.11.18-200.fc33.x86_64
* SELinux on?: I think so
* Any other system details we should know?:

**Additional context**

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,7 @@
+# 0.7.2 - 2021-05-28
+
+- Work with 0.9.x compiler
+
 # 0.7.1 - 2020-12-18
 
 - Allow CXX environment variable to override g++ for musl build (#500)

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "grpcio"
-version = "0.7.1"
+version = "0.7.2"
 edition = "2018"
 authors = ["The TiKV Project Developers"]
 license = "Apache-2.0"

--- a/compiler/src/util.rs
+++ b/compiler/src/util.rs
@@ -34,7 +34,7 @@ impl<'a> Iterator for NameSpliter<'a> {
         let mut meet_lower = false;
         for i in self.pos..self.name.len() {
             let c = self.name[i];
-            if b'A' <= c && c <= b'Z' {
+            if (b'A'..=b'Z').contains(&c) {
                 if meet_lower {
                     // So it should be AaA or aaA
                     pos = i;

--- a/grpc-sys/build.rs
+++ b/grpc-sys/build.rs
@@ -267,7 +267,7 @@ fn get_env(name: &str) -> Option<String> {
 
 // Generate the bindings to grpc C-core.
 // Try to disable the generation of platform-related bindings.
-fn bindgen_grpc(mut config: bindgen::Builder, file_path: &PathBuf) {
+fn bindgen_grpc(mut config: bindgen::Builder, file_path: &Path) {
     // Search header files with API interface
     let mut headers = Vec::new();
     for result in WalkDir::new(Path::new("./grpc/include")) {

--- a/interop/src/server.rs
+++ b/interop/src/server.rs
@@ -17,17 +17,6 @@ use grpc_proto::testing::messages::{
 use grpc_proto::testing::test_grpc::TestService;
 use grpc_proto::util;
 
-enum Error {
-    Grpc(grpc::Error),
-    Abort,
-}
-
-impl From<grpc::Error> for Error {
-    fn from(error: grpc::Error) -> Error {
-        Error::Grpc(error)
-    }
-}
-
 #[derive(Clone)]
 pub struct InteropTestService;
 
@@ -158,9 +147,10 @@ impl TestService for InteropTestService {
             sink.close().await?;
             Ok(())
         }
-        .map_err(|e: Error| match e {
-            Error::Grpc(grpc::Error::RemoteStopped) | Error::Abort => {}
-            Error::Grpc(e) => error!("failed to handle duplex call: {:?}", e),
+        .map_err(|e: grpc::Error| {
+            if !matches!(e, grpc::Error::RemoteStopped) {
+                error!("failed to handle duplex call: {:?}", e);
+            }
         })
         .map(|_| ());
         ctx.spawn(f)

--- a/src/call/client.rs
+++ b/src/call/client.rs
@@ -399,7 +399,7 @@ impl<Req> Sink<(Req, WriteFlags)> for StreamingCallSink<Req> {
             t.close_f = Some(close_f);
         }
 
-        if let Poll::Pending = Pin::new(t.close_f.as_mut().unwrap()).poll(cx)? {
+        if Pin::new(t.close_f.as_mut().unwrap()).poll(cx)?.is_pending() {
             // if call is finished, can return early here.
             call.check_alive()?;
             return Poll::Pending;

--- a/src/call/mod.rs
+++ b/src/call/mod.rs
@@ -533,7 +533,7 @@ impl StreamingBase {
         if !skip_finish_check {
             let mut finished = false;
             if let Some(close_f) = &mut self.close_f {
-                if let Poll::Ready(_) = Pin::new(close_f).poll(cx)? {
+                if Pin::new(close_f).poll(cx)?.is_ready() {
                     // Don't return immediately, there may be pending data.
                     finished = true;
                 }

--- a/src/call/mod.rs
+++ b/src/call/mod.rs
@@ -33,9 +33,9 @@ impl From<i32> for RpcStatusCode {
     }
 }
 
-impl Into<i32> for RpcStatusCode {
-    fn into(self) -> i32 {
-        self.0
+impl From<RpcStatusCode> for i32 {
+    fn from(c: RpcStatusCode) -> i32 {
+        c.0
     }
 }
 

--- a/src/cq.rs
+++ b/src/cq.rs
@@ -36,35 +36,39 @@ impl CompletionQueueHandle {
     }
 
     fn add_ref(&self) -> Result<()> {
+        let mut cnt = self.ref_cnt.load(Ordering::SeqCst);
         loop {
-            let cnt = self.ref_cnt.load(Ordering::SeqCst);
             if cnt <= 0 {
                 // `shutdown` has been called, reject any requests.
                 return Err(Error::QueueShutdown);
             }
             let new_cnt = cnt + 1;
-            if cnt
-                == self
-                    .ref_cnt
-                    .compare_and_swap(cnt, new_cnt, Ordering::SeqCst)
-            {
-                return Ok(());
+            match self.ref_cnt.compare_exchange_weak(
+                cnt,
+                new_cnt,
+                Ordering::SeqCst,
+                Ordering::SeqCst,
+            ) {
+                Ok(_) => return Ok(()),
+                Err(c) => cnt = c,
             }
         }
     }
 
     fn unref(&self) {
+        let mut cnt = self.ref_cnt.load(Ordering::SeqCst);
         let shutdown = loop {
-            let cnt = self.ref_cnt.load(Ordering::SeqCst);
             // If `shutdown` is not called, `cnt` > 0, so minus 1 to unref.
             // If `shutdown` is called, `cnt` < 0, so plus 1 to unref.
             let new_cnt = cnt - cnt.signum();
-            if cnt
-                == self
-                    .ref_cnt
-                    .compare_and_swap(cnt, new_cnt, Ordering::SeqCst)
-            {
-                break new_cnt == 0;
+            match self.ref_cnt.compare_exchange_weak(
+                cnt,
+                new_cnt,
+                Ordering::SeqCst,
+                Ordering::SeqCst,
+            ) {
+                Ok(_) => break new_cnt == 0,
+                Err(c) => cnt = c,
             }
         };
         if shutdown {
@@ -75,8 +79,8 @@ impl CompletionQueueHandle {
     }
 
     fn shutdown(&self) {
+        let mut cnt = self.ref_cnt.load(Ordering::SeqCst);
         let shutdown = loop {
-            let cnt = self.ref_cnt.load(Ordering::SeqCst);
             if cnt <= 0 {
                 // `shutdown` is called, skipped.
                 return;
@@ -85,12 +89,14 @@ impl CompletionQueueHandle {
             // Because `cnt` is initialized to 1, so minus 1 to make it reach
             // toward 0. That is `new_cnt = -(cnt - 1) = -cnt + 1`.
             let new_cnt = -cnt + 1;
-            if cnt
-                == self
-                    .ref_cnt
-                    .compare_and_swap(cnt, new_cnt, Ordering::SeqCst)
-            {
-                break new_cnt == 0;
+            match self.ref_cnt.compare_exchange_weak(
+                cnt,
+                new_cnt,
+                Ordering::SeqCst,
+                Ordering::SeqCst,
+            ) {
+                Ok(_) => break new_cnt == 0,
+                Err(c) => cnt = c,
             }
         };
         if shutdown {

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -80,3 +80,27 @@ pub use crate::security::{
 pub use crate::server::{
     CheckResult, Server, ServerBuilder, ServerChecker, Service, ServiceBuilder, ShutdownFuture,
 };
+
+/// A shortcut for implementing a service method by returning `UNIMPLEMENTED` status code.
+///
+/// Compiler will provide a default implementations for all methods to invoke this macro, so
+/// you usually won't call it directly. If you really need to, just call it like:
+/// ```ignored
+/// fn method(&self, ctx: grpcio::RpcContext, req: Request, resp: UnarySink<Response>) {
+///     unimplemented_call!(ctx, resp);
+/// }
+/// ```
+#[macro_export]
+macro_rules! unimplemented_call {
+    ($ctx:ident, $sink:ident) => {{
+        let f = async move {
+            let _ = $sink
+                .fail($crate::RpcStatus::new(
+                    $crate::RpcStatusCode::UNIMPLEMENTED,
+                    None,
+                ))
+                .await;
+        };
+        $ctx.spawn(f)
+    }};
+}

--- a/src/metadata.rs
+++ b/src/metadata.rs
@@ -16,11 +16,11 @@ fn normalize_key(key: &str, binary: bool) -> Result<Cow<'_, str>> {
     let mut is_upper_case = false;
     for b in key.as_bytes() {
         let b = *b;
-        if b >= b'A' && b <= b'Z' {
+        if (b'A'..=b'Z').contains(&b) {
             is_upper_case = true;
             continue;
-        } else if b >= b'a' && b <= b'z'
-            || b >= b'0' && b <= b'9'
+        } else if (b'a'..=b'z').contains(&b)
+            || (b'0'..=b'9').contains(&b)
             || b == b'_'
             || b == b'-'
             || b == b'.'
@@ -83,10 +83,10 @@ impl MetadataBuilder {
             }
         }
         let key = normalize_key(key, false)?;
-        self.add_metadata(&key, value.as_bytes())
+        Ok(self.add_metadata(&key, value.as_bytes()))
     }
 
-    fn add_metadata(&mut self, key: &str, value: &[u8]) -> Result<&mut MetadataBuilder> {
+    fn add_metadata(&mut self, key: &str, value: &[u8]) -> &mut MetadataBuilder {
         unsafe {
             grpc_sys::grpcwrap_metadata_array_add(
                 &mut self.arr.0,
@@ -96,7 +96,7 @@ impl MetadataBuilder {
                 value.len(),
             )
         }
-        Ok(self)
+        self
     }
 
     /// Add a metadata holding a binary value.
@@ -104,7 +104,7 @@ impl MetadataBuilder {
     /// `key` needs to have suffix (-bin) indicating a binary valued metadata entry.
     pub fn add_bytes(&mut self, key: &str, value: &[u8]) -> Result<&mut MetadataBuilder> {
         let key = normalize_key(key, true)?;
-        self.add_metadata(&key, value)
+        Ok(self.add_metadata(&key, value))
     }
 
     /// Create `Metadata` with configured entries.
@@ -221,7 +221,7 @@ impl Clone for Metadata {
         let mut builder = MetadataBuilder::with_capacity(self.len());
         for (k, v) in self.iter() {
             // use `add_metadata` to skip validation.
-            builder.add_metadata(k, v).unwrap();
+            builder.add_metadata(k, v);
         }
         builder.build()
     }

